relay-quotas/src/global.rs

```diff
@@ -0,0 +1,256 @@
+#[derive(Default)]
+pub struct GlobalRateLimits {
+    counters: RwLock<hashbrown::HashMap<BudgetKey, Arc<Mutex<SlottedCounter>>>>,
+}
+
+impl GlobalRateLimits {
+    /// Returns `true` if the global quota should be ratelimited.
+    ///
+    /// Certain errors can be resolved by syncing to redis, so in those cases
+    /// we try again to decrement the budget after syncing.
+    pub fn is_rate_limited(
+        &self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, RedisError> {
+        let Some(limit) = quota.limit else {
+            return Ok(false);
+        };
+
+        let key = BudgetKeyRef::new(quota);
+
+        let opt_val = self.counters.read().unwrap().get(&key).cloned();
+        let val = match opt_val {
+            Some(val) => val,
+            None => Arc::clone(self.counters.write().unwrap().entry_ref(&key).or_default()),
+        };
+
+        let mut lock = val.lock().unwrap();
+        lock.is_rate_limited(client, quota, quantity, limit)
+    }
+}
+
+/// Key for storing global quota-budgets locally.
+///
+/// Note: must not be used in redis. For that, use RedisQuota.key().
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct BudgetKey {
+    prefix: String,
+    window: u64,
+}
+
+impl From<&BudgetKeyRef<'_>> for BudgetKey {
+    fn from(value: &BudgetKeyRef<'_>) -> Self {
+        BudgetKey {
+            prefix: value.prefix.to_owned(),
+            window: value.window,
+        }
+    }
+}
+
+/// Used to look up a hashmap of [`BudgetKey`]-keys without a string allocation.
+///
+/// This works due to the 'Equivalent' trait in the hashbrown crate.
+#[derive(Clone, Copy, Hash)]
+struct BudgetKeyRef<'a> {
+    prefix: &'a str,
+    window: u64,
+}
+
+impl<'a> BudgetKeyRef<'a> {
+    fn new(quota: &'a RedisQuota) -> Self {
+        Self {
+            prefix: quota.prefix(),
+            window: quota.window(),
+        }
+    }
+}
+
+impl hashbrown::Equivalent<BudgetKey> for BudgetKeyRef<'_> {
+    fn equivalent(&self, key: &BudgetKey) -> bool {
+        self.prefix == key.prefix && self.window == key.window
+    }
+}
+
+struct SlottedCounter {
+    slot: usize,
+    counter: Counter,
+}
+
+impl SlottedCounter {
+    pub fn new() -> Self {
+        Self {
+            slot: 0,
+            counter: Counter::new(),
+        }
+    }
+
+    pub fn is_rate_limited(
+        &mut self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+        limit: u64,
+    ) -> Result<bool, RedisError> {
+        let current_slot = current_slot(quota.window());
+
+        match self.slot.cmp(&current_slot) {
+            Ordering::Greater => {
+                // TODO double check logic here
+                // in theory this should never happen only if someone messes with the system time
+                // be safe and dont rate limit
+                relay_log::error!("budget slot ahead of current slot");
```

Comment:
This is risky, if this actually happens this will happen for a longer period of time with a very high volume.

Ideally we would only log this error like once per second.